### PR TITLE
Implement `Extensions Negotiation` and `Worker Specific Hashrate Tracking` in applications

### PR DESCRIPTION
**Pull request by GitGab19**
This PR implements the two extensions on our applications.

It's based on my other work on `stratum`, which can be found here: https://github.com/stratum-mining/stratum/pull/1983.

Still WIP:
- [x] Pool
- [x] tProxy
- [x] JDC 
- [x] integration tests with extensions enabled

### CODE REVIEW

File: miner-apps/jd-client/src/lib/template_receiver/mod.rs
Comment (GitGab19) [inline miner-apps/jd-client/src/lib/template_receiver/mod.rs]: Same as https://github.com/stratum-mining/sv2-apps/pull/65/files#r2509734420

File: pool-apps/pool/src/lib/downstream/extensions_message_handler.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        );
+
+        // Get supported and required extensions from downstream data
+        let (supported_extensions, required_extensions) =
+            self.downstream_data.super_safe_lock(|data| {
+                (
+                    data.supported_extensions.clone(),
+                    data.required_extensions.clone(),
+                )
+            });
```
Comment (Shourya742) [inline pool-apps/pool/src/lib/downstream/extensions_message_handler.rs]: If we are not mutating this, lets put these fields directly inside the downstream struct

File: pool-apps/pool/src/lib/downstream/extensions_message_handler.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                "Downstream {}: Extension negotiation error: requested={:?}, supported={:?}, unsupported={:?}, missing_required={:?}",
+                self.downstream_id, requested, supported, unsupported, missing_required
+            );
+
+            let error = RequestExtensionsError {
+                request_id: msg.request_id,
+                unsupported_extensions: Seq064K::new(unsupported)
+                    .map_err(|_| PoolError::InvalidUnsupportedExtensionsSequence)?,
+                required_extensions: Seq064K::new(missing_required.clone())
+                    .map_err(|_| PoolError::InvalidRequiredExtensionsSequence)?,
```
Comment (Shourya742) [inline pool-apps/pool/src/lib/downstream/extensions_message_handler.rs]: lets add error context to these error. like `PoolError::InvalidRequiredExtensionsSequence(e)`

File: pool-apps/pool/src/lib/downstream/extensions_message_handler.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                .send(frame)
+                .await?;
+
+            // If required extensions are missing, the server SHOULD disconnect the client
+            if !missing_required.is_empty() {
+                error!(
+                    "Downstream {}: Client does not support required extensions {:?}. Server MUST disconnect.",
+                    self.downstream_id, missing_required
+                );
+                // TODO: Disconnect the client
```
Comment (Shourya742) [inline pool-apps/pool/src/lib/downstream/extensions_message_handler.rs]: You can return an error, and the downstream gonna get disconnected

File: miner-apps/translator/src/lib/sv2/channel_manager/channel_manager.rs [hunk 4]
```diff
@@ -525,16 +566,52 @@ impl ChannelManager {
...
+                                            && tlv.type_.field_type == TLV_FIELD_TYPE_USER_IDENTITY
+                                    })
+                                    .cloned()
+                            });
+                            let tlv_list = user_identity_tlv
+                                .map(|tlv| TlvList::from_slice(&[tlv]))
+                                .unwrap();
+                            let frame_bytes = tlv_list
+                                .build_frame_bytes_with_tlvs(Mining::SubmitSharesExtended(m))
+                                .unwrap();
```
Comment (Shourya742) [inline miner-apps/translator/src/lib/sv2/channel_manager/channel_manager.rs]: lets not use unwrap

File: miner-apps/jd-client/src/lib/upstream/mod.rs [hunk 5]
```diff
@@ -265,29 +311,23 @@ impl Upstream {
...
-    async fn handle_pool_message(&mut self) -> Result<(), JDCError> {
-        let mut sv2_frame = self.upstream_channel.upstream_receiver.recv().await?;
-
+    async fn handle_pool_message_frame(&mut self) -> Result<(), JDCError> {
         debug!("Received SV2 frame from upstream.");
-        let Some(message_type) = sv2_frame.get_header().map(|m| m.msg_type()) else {
-            return Ok(());
-        };
+        let mut sv2_frame = self.upstream_channel.upstream_receiver.recv().await?;
+        let header = sv2_frame.get_header().expect("header must be present");
```
Comment (Shourya742) [inline miner-apps/jd-client/src/lib/upstream/mod.rs]: propagate the error

File: miner-apps/jd-client/src/lib/config.rs [hunk 6]
```diff
@@ -164,6 +172,16 @@ impl JobDeclaratorClientConfig {
...
         self.share_batch_size
     }
+
+    /// Returns the supported extensions.
+    pub fn supported_extensions(&self) -> &Vec<u16> {
+        &self.supported_extensions
+    }
+
+    /// Returns the required extensions.
+    pub fn required_extensions(&self) -> &Vec<u16> {
```
Comment (Shourya742) [inline miner-apps/jd-client/src/lib/config.rs]:
```suggestion
    pub fn required_extensions(&self) -> &[u16] {
```

File: miner-apps/jd-client/src/lib/job_declarator/message_handler.rs [hunk 7]
```diff
@@ -15,10 +16,15 @@ use crate::{
 impl HandleCommonMessagesFromServerAsync for JobDeclarator {
     type Error = JDCError;
 
+    fn get_negotiated_extensions_with_server(&self, _server_id: Option<usize>) -> Vec<u16> {
+        vec![]
```
[the remark this answers is not in the document]
Comment (GitGab19) [reply miner-apps/jd-client/src/lib/job_declarator/message_handler.rs]: It's not a concern in this specific case because we don't have any extension which adds TLV fields to any JD message, but my point still remains the same.

File: pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs [hunk 8]
```diff
@@ -682,6 +724,10 @@ impl HandleMiningMessagesFromClientAsync for ChannelManager {
                     return Ok(vec![(downstream_id, Mining::SubmitSharesError(error)).into()]);
                 };
 
+                if let Some(user_identity) = user_identity {
+                    extended_channel.set_user_identity(user_identity.as_string_or_hex());
+                }
```
Comment (plebhash) [inline pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs]: I worry that this is going in a bad direction

we want to associate shares with `user_identity`, while here we're simply changing the `user_identity` that's associated with the channel every time there's a new share, without any meaningful effect over how we actually track hashrate against different `user_identity`s

File: miner-apps/jd-client/src/lib/channel_manager/downstream_message_handler.rs [hunk 9]
```diff
@@ -1220,7 +1272,10 @@ impl HandleMiningMessagesFromClientAsync for ChannelManager {
                                     prev_hash: prev_hash.prev_hash.clone(),
                                 };
                                 messages.push(JobDeclaration::PushSolution(push_solution.clone()).into());
-                                messages.push(Mining::SubmitSharesExtended(upstream_message.into_static()).into());
+                                // TODO here we should add the user_identity TLV to the SubmitSharesExtended frame
```
Comment (plebhash) [inline miner-apps/jd-client/src/lib/channel_manager/downstream_message_handler.rs]: will this be done on a follow up PR?

File: pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs [hunk 10]
```diff
@@ -45,10 +48,28 @@ impl HandleMiningMessagesFromClientAsync for ChannelManager {
...
+            .super_safe_lock(|channel_manager_data| {
+                channel_manager_data
+                    .downstream
+                    .get(&downstream_id)
+                    .map(|downstream| {
+                        downstream
+                            .downstream_data
+                            .super_safe_lock(|data| data.negotiated_extensions.clone())
+                    })
+                    .expect("negotiated_extensions must be present")
```
[the remark this answers is not in the document]
Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs]: I feel the method is not necessary here, it would only make things more convoluted IMO.

**Review by plebhash — APPROVED** [review]

**Review by xyephy — COMMENTED** [review]
@GitGab19 Nice work on the extensions implementation! Everything compiles, tests pass, and all of Shourya's feedback is addressed. 

 Edge Case Found (not blocking, pre-existing from main):
Line 423 in `channel_manager.rs` can panic if downstream sends binary (non-UTF8) `user_identity`:

- Binary input (255 bytes) → hex encoding (0x + 510 chars = 512 bytes)
- Append `.translator-proxy` (+18 bytes) = 530 bytes total
- `.try_into::<Str0255>().unwrap()` panics because 530 > 255 max

This is unrelated to your extensions work - just something I noticed during review. Happy to open a follow-up PR if you think it's worth fixing, otherwise we can leave it. What do you think?

File: miner-apps/translator/src/lib/sv2/upstream/upstream.rs [hunk 11]
```diff
@@ -249,19 +255,41 @@ impl Upstream {
...
 
-        self.handle_common_message_frame_from_server(None, message_type, payload)
+        self.handle_common_message_frame_from_server(None, header, payload)
             .await?;
         debug!("Upstream: handshake completed successfully.");
+
+        // Send RequestExtensions message if there are any required extensions
+        if !self.required_extensions.is_empty() {
+            let require_extensions = RequestExtensions {
+                request_id: 0,
```
Comment (Shourya742) [inline miner-apps/translator/src/lib/sv2/upstream/upstream.rs]: lets make this to 1

File: pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs [hunk 12]
```diff
@@ -45,10 +48,28 @@ impl HandleMiningMessagesFromClientAsync for ChannelManager {
         Ok(true)
     }
 
+    fn get_negotiated_extensions_with_client(&self, client_id: Option<usize>) -> Vec<u16> {
```
[the remark this answers is not in the document]
Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs]: The method is defined in the trait, so I cannot change it here.

If you're worried about the expect(s), that's a different discussion.

The first expect() is the same we have in every other methods impl, while the second one is safe because it can never happen that a downstream doesn't have the `negotiated_extensions`.

[the remark this answers is not in the document]
Comment (Shourya742) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs]: Those methods don't change dynamically during runtime, `get_negotiated_extensions_with_client` does if extensions are enabled.

[the remark this answers is not in the document]
Comment (GitGab19) [reply pool-apps/pool/src/lib/channel_manager/mining_message_handler.rs]: Adapted all the apps to the changes made on https://github.com/stratum-mining/stratum/pull/1983.
